### ISSUE TITLE
Public-facing API for Styled Text Output

### ISSUE THREAD

**Issue by JacisNonsense** [gradle/gradle#5825]
### Expected Behavior
Some form of public API similar to StyledTextOutput and its associated Factory

### Current Behavior
Must rely on internal APIs, as StyledTextOutput being public was initially a class leakage problem. Now that it's all internal, no such public API exists.

### Context
Non-fatal warnings, or notices for unstable versions in the console can easily be lost in the sea of text. In my particular case, I have a task that prints a bunch of self-tests on the gradle project to help diagnose configuration issues, and I'd like to highlight potential problems in red without failing the build.

**Comment by rpalcolea** [comment]
Hi folks,

Are there any plans to make this public soon? It will be very helpful for plugin authors

**Comment by NielsDoucet** [comment]
I agree. Drawing attention to warnings or errors produced during a build is critical to make users notice potential mistakes. Having coloured output would help a lot here.
We have some may's and should's in our guidelines which shouldn't break a build if not respected, but they could indicate a way to improve build stability and performance.

**Comment by melix** [comment]
I have been invited to give my opinion on this, and I'm speaking as an individual, not in the name of the company.

I agree we should provide an API for plugin authors, or build script authors, to provide rich outputs. At the same time, I personally find the internal "StyledOutput" incredibly complex and unpleasant to use.

This is why, while implementing better variant-aware dependency management errors I experimented with an alternate approach, based on good old tagged text.

Do you want styled output? Then what you can write is something like:

`This is an <failure>error</failure> and this is a <info>simple text</info>`

Then Gradle takes care of rendering it properly depending on the attached console. If it's a rich console, it can use the rich color output. Otherwise, this is just convertible to plain text.

I find this approach much more usable for such a simple use case. One way to move forward would just be to provide a new logger, either through `logger.styled()` or `getStyledLogger()`, so that you could just write:

```
logger.styled().error("This is an <failure>error</failure> and this is a <info>simple text</info>")
```

Just my 2 cents.

**Comment by rpalcolea** [comment]
Hey @melix,

I like your suggestions. Tagging is great! It makes sense to just highlight the things that matter and not the whole text. Also, agree with the complexity behind `StyledTextOutput`.

Hopefully folks could use this wisely in plugins if added to gradle. Otherwise we will end with such nice Christmas trees in the logs 😅

**Comment by rpalcolea** [comment]
Please keep this open

**Comment by lennartj** [comment]
## Standardized styled output: Still required and desired

The need to provide styled output from plugins is still very much present and - unless I have missed the documentation on it - not adressed or standardized by Gradle for plugin developers. The StyledTextOutput and related classes are defined in the **internal** gradle structure, so not really viable as a stable option for plugin developers.

```java
package org.gradle.internal.logging.text;

/**
 * Provides streaming of styled text, that is, a stream of text with inline styling information. Implementations are not
 * required to be thread-safe.
 */
public interface StyledTextOutput extends Appendable {
```

Moreover, the ability to reliably test styled output in the plugin test frameworks is also missing. (Again, unless I have missed finding some documentation on it).

- Could you please point us to some guidance on this?
- What are the implications of inserting - say - <info>Foobar</info> in a gradle logger? Will it **always** be styled if the targeting console has colored capability? How can one understand the configuration (i.e. how "info" will be converted to a "green" color output)?
